### PR TITLE
Implement Retrofit using Repository Pattern and Dependency Injection

### PR DESCRIPTION
**Pull request by AlexVDV116**
This pull request introduces the Repository pattern to enhance data management in the application and abstract direct interaction with the REST API service away from the ViewModel. It involves the use of Retrofit for network operations and dependency injection for improved testability. Additionally, unit tests have been created for both the ViewModel and the Repository to ensure the robustness of the implemented features.

**Changes made:**

- [X] Seperate the UI layer and the Data layer using the repository pattern
Created the `UserRepository` interface and `NetworkUserRepository` that implements this interface that serve as an intermediary layer between the data sources (network, local database) and the UI components.

- [X] Creata an Application container
Acts as a centralized container for managing dependencies related to network requests and data retrieval. It follows principles of dependency injection, lazy initialization, and separation of concerns.

- [X] Add the userRepository to the userViewModel using the Factory Pattern
Because the Android framework does not allow a ViewModel to be passed values in the constructor when created, we implement a `ViewModelProvider.Factory` object, which lets us get around this limitation. The Factory pattern is a creational pattern used to create objects. The `UserViewModel.Factory` object uses the application container to retrieve the `userRepository`, and then passes this repository to the ViewModel when the ViewModel object is created.

- [X] Create fake data and dependencies for tests
By using interfaces and subsequently using dependency injection to include classes that inherit from those interfaces, we simulate the functionality of those dependencies using fake classes made solely for the purpose of testing. Injecting fake classes and data sources for testing allows code to be tested in isolation, with repeatability and consistency.

- [X] Created TestDispatcherRule for coroutine testing
Since the Main dispatcher is only available in a UI context, you must replace it with a unit-test-friendly dispatcher. The Kotlin Coroutines library provides a coroutine dispatcher for this purpose called `TestDispatcher`. The `TestDispatcher` needs to be used instead of the Main dispatcher for any unit test in which a new coroutine is made, as is the case with the `getUsers() `function from the view model.

- [X] Created unit tests for UserViewModel and UserRepository
By passing the fake API service, any calls to the `RmcApiService` property in the repository result in a call to the `FakeRmcApiService`. By passing fake classes for dependencies, we control exactly what the dependency returns. This approach ensures that the code we are testing doesn't depend on untested code or APIs that could change or have unforeseen problem.

**Future work:**

- [X] Establish connection with the `rmc-api` API on localhost.
- [x] Implement connection with the Room library.
- [ ] Extract data from the API into composables/screens.

### CODE REVIEW

File: app/src/main/java/com/digitalarchitects/rmc_app/network/RmcApiService.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+import com.digitalarchitects.rmc_app.model.User
+import com.digitalarchitects.rmc_app.model.Vehicle
+import retrofit2.http.GET
+
+
+interface RmcApiService {
+    /**
+     * Returns a [List] of [User] and this method can be called from a Coroutine.
+     * The @GET annotation indicates that the "photos" endpoint will be requested with the GET
```
Comment (vderlinden) [inline app/src/main/java/com/digitalarchitects/rmc_app/network/RmcApiService.kt]: "photos"? Do you mean Users

File: app/src/main/java/com/digitalarchitects/rmc_app/network/RmcApiService.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+     * HTTP method
+     */
+    @GET("user/users")
+    suspend fun getUsers(): List<User>
+
+    /**
+     * Returns a [List] of [Vehicle] and this method can be called from a Coroutine.
+     * The @GET annotation indicates that the "photos" endpoint will be requested with the GET
```
Comment (vderlinden) [inline app/src/main/java/com/digitalarchitects/rmc_app/network/RmcApiService.kt:19]: "photos"? Do you mean Users

**Review by vderlinden — COMMENTED** [review]
@AlexVDV116 Looks good so far, and thank you for the detailed explanation!

File: app/src/main/java/com/digitalarchitects/rmc_app/network/RmcApiService.kt (hunk 1 above)
Comment (AlexVDV116) [reply app/src/main/java/com/digitalarchitects/rmc_app/network/RmcApiService.kt]: Yes I meant users, comment is copied and pasted from one of the Android Developers codelab. 😄

**Comment by AlexVDV116** [comment]
**Changes made:**

- [X] Establish connection with the `rmc-api` API on localhost.

**Managed to establish a connection with the `rmc-api` using the Retrofit library:**
- [Add a Network Security Configuration File to configure a custom CA](https://developer.android.com/privacy-and-security/security-config#manifest)
- [Configure a custom CA so we can connect to a host with a self-signed certificate](https://developer.android.com/privacy-and-security/security-config#ConfigCustom)
- Configured the retrofit `BASE_URL` so it connects with the localhost using HTTPS
- Added the Kotlinx Serialization plugin for the Retrofit Convertor Factory so it can serialize and deserialize objects
- Configured the `RmcApiService` interface getUsers request to implement the Authorization header
- Configured the `userRepository` `getUsers()` fun to send a JWT Bearer token in the Authorization header
- Created the `RmcApiTestScreen` for debugging requests to the `rmc-api` host

**NOTE:** Had to update Kotlin version, KotlinCompilerExtensionVersion and Compose due to some compatibility issues:
[Used the following graph to tune different versions](https://developer.android.com/jetpack/androidx/releases/compose-kotlin#pre-release_kotlin_compatibility)

**Future work:**

- [ ] Implement connection with the Room library.
- [ ] Extract data from the API into composables/screens.

File: app/build.gradle.kts
Comment (Brian174tech) [inline app/build.gradle.kts:1]: Is de kotlinCOmpilerExtension version upgrade nodig van 1.4.3 naar 1.5.6. Ik weet namelijk niet of dit wel compatable is met Room sraks

File: app/src/main/java/com/digitalarchitects/rmc_app/data/UserViewModel.kt
Comment (Brian174tech) [inline app/src/main/java/com/digitalarchitects/rmc_app/data/UserViewModel.kt:1]: die copy paste inclusief copyright shit uit de codelabs moet wel weg...
Sommige comments zijn misschien ook niet nodig. (maar geen prio)

File: app/src/main/java/com/digitalarchitects/rmc_app/model/User.kt
Comment (Brian174tech) [inline app/src/main/java/com/digitalarchitects/rmc_app/model/User.kt:1]: Eentje om mee te nemen is wel de benam ing van de klasses, we hebben straks heel veel dezelfde klassenamen. misschien specificer(was meer een algemene opmerking, zit zelf ook te stoeien met de Room tabel benamingen)

File: app/src/main/res/drawable/ic_connection_error.xml
Comment (Brian174tech) [inline app/src/main/res/drawable/ic_connection_error.xml:1]: wederom licensing en copyright, weg ermee

File: app/src/main/res/drawable/loading_img.xml
Comment (Brian174tech) [inline app/src/main/res/drawable/loading_img.xml:1]: wat zijn dit?

File: app/build.gradle.kts
Comment (Brian174tech) [inline app/build.gradle.kts:1]: Is de kotlinCOmpilerExtension version upgrade nodig van 1.4.3 naar 1.5.6. Ik weet namelijk niet of dit wel compatible is met Room straks

File: app/src/main/java/com/digitalarchitects/rmc_app/data/UserViewModel.kt
Comment (Brian174tech) [inline app/src/main/java/com/digitalarchitects/rmc_app/data/UserViewModel.kt:1]: die copy paste inclusief copyright shit uit de codelabs moet wel weg...
Sommige comments zijn misschien ook niet nodig. (maar geen prio)

File: app/src/main/java/com/digitalarchitects/rmc_app/model/User.kt
Comment (Brian174tech) [inline app/src/main/java/com/digitalarchitects/rmc_app/model/User.kt:1]: Eentje om mee te nemen is wel de benam ing van de klasses, we hebben straks heel veel dezelfde klassenamen. misschien specifiekere benamingen maken(was meer een algemene opmerking, zit zelf ook te stoeien met de Room tabel benamingen)

File: app/src/main/res/drawable/ic_connection_error.xml
Comment (Brian174tech) [inline app/src/main/res/drawable/ic_connection_error.xml:1]: wederom licensing en copyright, weg ermee

File: app/src/main/res/drawable/loading_img.xml
Comment (Brian174tech) [inline app/src/main/res/drawable/loading_img.xml:1]: wat zijn dit?

File: app/src/main/res/xml/network_security_config.xml
Comment (Brian174tech) [inline app/src/main/res/xml/network_security_config.xml:1]: wat is dit?

File: app/src/test/java/com/digitalarchitects/rmc_app/fake/FakeDataSource.kt
Comment (Brian174tech) [inline app/src/test/java/com/digitalarchitects/rmc_app/fake/FakeDataSource.kt:1]: beetje vergelijkbaar met de dto's die ik had gemaakt, dummy data. misschien dat we beter het een of het ander aanhouden. Het is wel fijn als je data met properties aanhoudt zoals deze in de API staan, er missen nu properties en de repository data werkt nu niet met bv drawables.

File: app/src/test/java/com/digitalarchitects/rmc_app/fake/FakeNetworkUserRepository.kt
Comment (Brian174tech) [inline app/src/test/java/com/digitalarchitects/rmc_app/fake/FakeNetworkUserRepository.kt:1]: heb je echt coroutines/suspend functies nodig voor een fake repository?

File: build.gradle.kts
Comment (Brian174tech) [inline build.gradle.kts:1]: Is deze update ook nodig? Zo niet dan zou ik ect even wachten met tussentijds dependencies veranderen terwijl we veel banches paralel hebben lopen...

**Review by Brian174tech — APPROVED** [review]
Heb best wat comments geleverd zo, hoop dat je ze niet te persoonlijk opvat en ze kan beantwoorden voor mij. Begrijp uit de whatsapp dat je alweer een stuk verder bent en ziet er netjes uit - dus ga er van uit dat het allemaal werkt en niet alles testen.(Je hebt zelf ook getest en Paul heeft hem ook gedraaid begrijp ik. Dus dat zit sws wel goed) 
Was weer een hele kluif denk ik zo, goed werk kerel.

File: app/build.gradle.kts
Comment (AlexVDV116) [reply app/build.gradle.kts:1]: Lijkt me dat Room ook compatible is met de laatste verise van KotlinCompiler.

File: app/src/main/java/com/digitalarchitects/rmc_app/data/UserViewModel.kt
Comment (AlexVDV116) [reply app/src/main/java/com/digitalarchitects/rmc_app/data/UserViewModel.kt:1]: Mee eens, tijdens development wel handig zodat jullie ook weten wat deze klasse doet. Met oplevering kunnen nog eens alles nalopen en controleren wat nog nodig is.

File: app/src/main/java/com/digitalarchitects/rmc_app/model/User.kt
Comment (AlexVDV116) [reply app/src/main/java/com/digitalarchitects/rmc_app/model/User.kt:1]: Welke klasse namen heb je het dan over? De data klasse naam voor User, Vehicle and Rental lijken me vrij logisch. Als dit interfereert met klasse name in Room moeten we hier even naar kijken wat best practice is.

File: app/src/main/res/drawable/ic_connection_error.xml
Comment (AlexVDV116) [reply app/src/main/res/drawable/ic_connection_error.xml:1]: Done.

File: app/src/main/res/drawable/loading_img.xml
Comment (AlexVDV116) [reply app/src/main/res/drawable/loading_img.xml:1]: Images gebruikt in de RmcApiSreen voor de verschillende uiStates.  De RmcApiTestScreen kan 3 composables weergeven afhankelijk van de UserUiState in de UserViewModel.

In de fun getUsers() van de UserViewModel zie je een try catch. Afhankelijk van de response van de Retrofit request naar de API wordt een verschillend composable getoond, die gebruik maken van de 2 drawables die ik heb toegevoegd.

Dit gebruik ik op het moment alleen om te debuggen, heel de RmcApiTestScreen kan wanneer alles werkt weer weg.

File: app/src/main/res/xml/network_security_config.xml
Comment (AlexVDV116) [reply app/src/main/res/xml/network_security_config.xml:1]: Wederom string resources gebruikt in de RmcApiTestScreen voor debugging purposes. Tijdelijk en kan wanneer ik deze branch gereed heb weg.

File: app/src/test/java/com/digitalarchitects/rmc_app/fake/FakeDataSource.kt
Comment (AlexVDV116) [reply app/src/test/java/com/digitalarchitects/rmc_app/fake/FakeDataSource.kt:1]: Ja de data classes in het Android project waren niet gelijk aan die in de API. Heb dit nu rechtgetrokken in een nieuwe commit.

File: app/src/test/java/com/digitalarchitects/rmc_app/fake/FakeNetworkUserRepository.kt
Comment (AlexVDV116) [reply app/src/test/java/com/digitalarchitects/rmc_app/fake/FakeNetworkUserRepository.kt:1]: Ja. 

Dit is een Unit test die de UserRepository test. Deze klasse implementeerd de UserRepository interface. Bij een unit test willen we enkel en alleen de functionaliteit van de unit testen, in dit geval de UserRepository. In de andere test de UserViewModel. 

Omdat onze klasses (Repository, ViewModel en ApiService) afhankelijk zijn van elkaar en we gebruik maken van dependency injection maken we gebruik van FakeNetworkRepository en FakeRmcApiService en FakeDataSource om de functionaliteit van deze dependencies simuleren.

Ik citeer:

> Each local test needs to only test one thing. For example, when you test the functionality of the view model, you do not want to test the functionality of the repository or the API service. Likewise, when you test the repository, you do not want to test the API service.

> By using interfaces and subsequently using dependency injection to include classes that inherit from those interfaces, you can simulate the functionality of those dependencies using fake classes made solely for the purpose of testing. Injecting fake classes and data sources for testing allows code to be tested in isolation, with repeatability and consistency.

File: build.gradle.kts
Comment (AlexVDV116) [reply build.gradle.kts:1]: Ja dit is nodig. Voor Retrofit functionaliteit moest de Compose Compiler Version geüpgraded worden. Conform de [Kotlin Compatibility Map](https://developer.android.com/jetpack/androidx/releases/compose-kotlin) moet de Kotlin versie ook worden geüpgraded build errors te voorkomen.

File: app/src/test/java/com/digitalarchitects/rmc_app/fake/FakeNetworkUserRepository.kt
Comment (Brian174tech) [reply app/src/test/java/com/digitalarchitects/rmc_app/fake/FakeNetworkUserRepository.kt:1]: ja nee je hebt gelijk, had even niet door dat het een test was

**Review by Brian174tech — APPROVED** [review]
Je updates gezien en behandeld, ziet er mooi uit. Dankjewel.
Zoals net besproken via de whatsapp alle branches de master in, een apart resolve (branch/merge voor alle conflicts) maken en dan kunnen we met zijn drie weer schoon beginnen!

**Pull request merged**